### PR TITLE
Use .class files instead of compiling the sources

### PR DESCRIPTION
**Pull request by JaroslavTulach**
The goal is to give certain project types, or their configurations, a chance to use existing `.class` files rather than compiling from `.java`. The primary motivation for this API is to support following usecase with [mx](http://github.com/graalvm/mx) based [graal](http://github.com/oracle/graal) sources:
```bash
$ cd graal/truffle
$ mx build
$ netbeans --open .
```
The build command compiles everything and places the classes into `bin` directories. By using the API introduced in this pull request, the project can instruct the NetBeans Java support to copy those `.class` files and thus speed the initialization up.

The support shall be generally useful in all project types. Yet, it is up to individual projects to turn it on. There has been pushback against this copying approach in the past, so we definitely want opt-in behavior, the default should remain as it is.

### CODE REVIEW

**Comment by tzezula** [comment]
Maybe it's good to mention that preferring class files to source files has some limitations.
The find usages (refactoring) will not work on source level annotations, static fields initialized by compilation time constant, etc.
I am not saying that's a big problem but should be known.

File: ide/api.java.classpath/src/org/netbeans/spi/java/queries/BinaryForSourceQueryImplementation2.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+ * {@codesnippet org.netbeans.api.java.queries.BinaryForSourceQuery2Test.SampleQuery}
+ * 
+ * @param <Result> any type this implementation wants to use as a result
+ * @see BinaryForSourceQuery
+ * @see SourceForBinaryQuery
+ * @see BinaryForSourceQueryImplementation
+ * @see SourceForBinaryQueryImplementation
+ * @since 1.58
+ */
+public interface BinaryForSourceQueryImplementation2<Result> extends BinaryForSourceQueryImplementation {
```
Comment (tzezula) [inline ide/api.java.classpath/src/org/netbeans/spi/java/queries/BinaryForSourceQueryImplementation2.java]: 👍🏼 Nicely documented.

File: java/java.source.base/src/org/netbeans/modules/java/source/indexing/JavaCustomIndexer.java [hunk 2]
```diff
@@ -147,6 +151,76 @@ protected void index(final Iterable<? extends Indexable> files, final Context co
                 JavaIndex.LOG.fine("Ignoring request with no root"); //NOI18N
                 return;
             }
+            BinaryForSourceQuery.Result2 binRes = BinaryForSourceQuery.findBinaryRoots2(root.toURL());
```
Comment (tzezula) [inline java/java.source.base/src/org/netbeans/modules/java/source/indexing/JavaCustomIndexer.java]: There is an additional call to project query but should not be a big problem as it's followed by query for bootpath, compilepath, sourcepath, source level and encoding. Also it's O(n) to roots.

File: java/java.source.base/src/org/netbeans/modules/java/source/indexing/JavaCustomIndexer.java [hunk 3]
```diff
@@ -147,6 +151,76 @@ protected void index(final Iterable<? extends Indexable> files, final Context co
...
+                        long modified = ch.lastModified().getTime();
+                        if (newestFile == null || newestFile < modified) {
+                            newestFile = modified;
+                        }
+                    }
+                }
+
+                boolean binariesAreNewer = true;
+                for (Indexable index : files) {
+                    if (!binariesAreNewer) {
```
Comment (tzezula) [inline java/java.source.base/src/org/netbeans/modules/java/source/indexing/JavaCustomIndexer.java]: Can be removed, no?
The `binariesAreNewer = true` is followed by `break`.

File: java/java.source.base/src/org/netbeans/modules/java/source/indexing/JavaCustomIndexer.java [hunk 4]
```diff
@@ -147,6 +151,76 @@ protected void index(final Iterable<? extends Indexable> files, final Context co
...
+                                name = ch.getName() + ".sig";
+                            }
+                            File toDir = new File(copyTo, path.replace('/', File.separatorChar));
+                            toDir.mkdirs();
+                            File to = new File(toDir, name);
+                            try (OutputStream os = new FileOutputStream(to); InputStream is = ch.getInputStream()) {
+                                FileUtil.copy(is, os);
+                            }
+                        }
+                    }
```
Comment (tzezula) [inline java/java.source.base/src/org/netbeans/modules/java/source/indexing/JavaCustomIndexer.java]: I think that `APTUtils.sourceRootRegistered(context.getRoot(), context.getRootURI());` should be called here.

**Comment by tzezula** [comment]
I am not sure how this will behave when the project is using annotation processors generating sources. Did you try that?
Thanks

**Comment by JaroslavTulach** [comment]
> Maybe it's good to mention that preferring class files to source files has some limitations.

Right. Another reason for the support to be opt-in only.

> The find usages (refactoring) will not work on source level annotations, static fields initialized by compilation time constant, etc.

Right.

> Another limitation comes from annotation processing. There will be no sources generated by AP only binaries.

In the `mx` based projects, I can put the `src_gen` folder on some classpath. E.g. the sources will be there, it is just a question whether IDE finds them... I will need to test this change on the real `mx` projects before merging it to make sure it really works.

**Comment by jlahoda** [comment]
Regarding APs, the tricky aspect might be when the APs need to read the sources from previous compilation iterations in incremental compilation. It may even be necessary to copy the sources as well.

I appreciate that the API was adjusted to a more common structure.

Overall, I don't have problem with this patch.

File: java/java.source.base/src/org/netbeans/modules/java/source/indexing/JavaCustomIndexer.java (hunk 3 above)
Comment (JaroslavTulach) [reply java/java.source.base/src/org/netbeans/modules/java/source/indexing/JavaCustomIndexer.java]: Done in 64160669b763

File: java/java.source.base/src/org/netbeans/modules/java/source/indexing/JavaCustomIndexer.java (hunk 4 above)
Comment (JaroslavTulach) [reply java/java.source.base/src/org/netbeans/modules/java/source/indexing/JavaCustomIndexer.java]: Is c1dc05a8138b correct?

**Comment by JaroslavTulach** [comment]
Here are the changes generated by `ant -f ide/api.java.classpath check-sigtest -Dsigtest.check.type=strictcheck`:
```
  [sigtest] Added Classes
  [sigtest] -------------
  [sigtest] 
  [sigtest] org.netbeans.api.java.queries.BinaryForSourceQuery$Result2
  [sigtest] org.netbeans.spi.java.queries.BinaryForSourceQueryImplementation2
  [sigtest] 
  [sigtest] Added Nested Classes
  [sigtest] --------------------
  [sigtest] 
  [sigtest] org.netbeans.api.java.queries.BinaryForSourceQuery:         nested public abstract static org.netbeans.api.java.queries.BinaryForSourceQuery$Result2
  [sigtest] 
  [sigtest] Added Methods
  [sigtest] -------------
  [sigtest] 
  [sigtest] org.netbeans.api.java.queries.BinaryForSourceQuery:         method public final static org.netbeans.api.java.queries.BinaryForSourceQuery$Result2 org.netbeans.api.java.queries.BinaryForSourceQuery.findBinaryRoots2(java.net.URL)
  [sigtest] 
```
I'd like to integrate the change into NetBeans 11. @lkishalmi, can you cherry pick it once it gets into master?

**Comment by tzezula** [comment]
Seems OK to me.

File: platform/libs.osgi/external/osgi-7.0-license.txt [hunk 5]
```diff
@@ -1,6 +1,6 @@
 Name: OSGi
 Version: 7.0
-Files: osgi.core-7.0.0.jar org.osgi.compendium-5.0.0.jar
+Files: osgi.core-7.0.0.jar osgi.cmpn-7.0.0.jar
```
Comment (JaroslavTulach) [inline platform/libs.osgi/external/osgi-7.0-license.txt]: Testing this change in [OracleLabs IGV](https://www.graalvm.org/docs/graalvm-as-a-platform/implement-language/#igv) revealed that the `org.osgi.service.resolver.ResolveContext` is present twice:
* in the OSGi core JAR
* in the OSGi companion JAR

when the companion version took preference, the Felix boot failed due to linkage error as `onCancel(Runnable)` method was missing. Updating to `7.0.0` (e.g. the same version as the OSGi core) should solve this inconsistency.

**Comment by JaroslavTulach** [comment]
I did testing on IGV and things seem to work relatively well. Let's integrated. @lkishalmi, please consider backporting this into `release110` branch. That would make it easier for some future version of IGV to use the new API.

**Pull request merged**